kiwi/bootloader/config/grub2.py

```diff
@@ -233,14 +233,14 @@ def setup_disk_image_config(
                 config_file.replace(self.root_mount.mountpoint, '')
             ]
         )
-        if self.root_filesystem_is_overlay:
+        if self.root_filesystem_is_overlay or Defaults.is_buildservice_worker():
             # grub2-mkconfig has no idea how the correct root= setup is
-            # for disk images created with overlayroot enabled. Because
-            # of that the mkconfig tool just finds the raw partition loop
-            # device and includes it which is wrong. In this particular
-            # case we have to patch the written config file and replace
-            # the wrong root= reference with the correct value such that
-            # the dracut-kiwi-overlay module can boot the system.
+            # for disk images created with overlayroot enabled or in a
+            # buildservice worker environment. Because of that the mkconfig
+            # tool just finds the raw partition loop device and includes it
+            # which is wrong. In this particular case we have to patch the
+            # written config file and replace the wrong root= reference with
+            # the correct value.
```

Comment:
> And what if we just always set the root= parameter from KIWI and that's it?

we had that, it lead to double entries because grub2-mkconfig writes their own root= parameter

> I believe that we can pass GRUB_DEVICE_UUID as en environment variable to grub2-mkconfig

this does not work from my perspective. grub2-mkconfig always creates their own root= setup except for NFSROOT. The variable GRUB_DEVICE_UUID is written by the tooling. In addition every distro has their own magic set of patches to that shell code

Actually I liked the way we have it now, it's explicit and only touches the output for the overlay case which was not supported by grub2-mkconfig and probably never will and we extend that to obs workers which are special too and outside any OS standard

Reply:
in any other case we trust grub2-mkconfig and expect it to do the right thing without touching any env setup by us. That's how it should work, except for the exceptions

Reply:
there is no real API on the grub2 tools. Thus I'm not sure why you think using one of the env variables would lead to a cleaner solution :)

Reply:
We can also test if the patching code does the right thing, using a shell variable cannot be tested. All kinds of weird things can happen depending on how grub2-mkconfig works with it

Reply:
```
if [ "x${NFSROOT_DEVICE}" != "x" ]; then
    GRUB_DEVICE=""
    GRUB_DEVICE_UUID=""
    GRUB_FS="unknown"
else
    # Device containing our userland.  Typically used for root= parameter.
    GRUB_DEVICE="`${grub_probe} --target=device /`"
    GRUB_DEVICE_UUID="`${grub_probe} --device ${GRUB_DEVICE} --target=fs_uuid 2> /dev/null`" || true


    if [ x"$GRUB_FS" = x ] || [ x"$GRUB_FS" = xunknown ]; then
    GRUB_FS="$(stat -f --printf=%T / || echo unknown)"
    fi
fi
```

says everything

Reply:
and all this code is not consistent across distributions

Reply:
> and all this code is not consistent across distributions

Well yes this is the major concern here I guess. 

I came to this idea because of 
```bash
# These are defined in this script, export them here so that user can
# override them.
export GRUB_DEVICE \
  GRUB_DEVICE_UUID \
  GRUB_DEVICE_BOOT \
  GRUB_DEVICE_BOOT_UUID \
  GRUB_FS \
  GRUB_FONT \
  GRUB_PRELOAD_MODULES \
  GRUB_ACTUAL_DEFAULT
```
I found it in the code upstream (https://git.savannah.gnu.org/git/grub.git) and also in our SUSE packages. However I agree that the amount of patches that grub2 is having is a real concern.

> there is no real API on the grub2 tools. Thus I'm not sure why you think using one of the env variables would lead to a cleaner solution :)

Well I was just wondering on the way to tell the UUID of the device to the grub2-mkconfig. So no later changes are required neither any specific case consideration.

My concern is more generic in the spirit that I dislike adding OBS specific code in KIWI. We already have some, I'd love to keep it to the minimum and reduce it whenever possible.

Reply:
> Actually I liked the way we have it now, it's explicit and only touches the output for the overlay case which was not supported by grub2-mkconfig and probably never will and we extend that to obs workers which are special too and outside any OS standard

Well this is true, the only cases that are failing are really specific and exceptional. I agree on that.

Ok, I can live with the code of this PR, as it is really not very intrusive and simple. I understand your concerns regarding grub2-mkconfig consistency across distros. Approving the PR.

Another option to determine if the patch is required could be the gurb2-probe to validate if grub2-mkconfig is capable to guess the device_uuid and if not, just manually set the `root=`.

Reply:
I completely agree to your concern. I just don't see a good solution since an obs worker seems to be special at several places and still designed to build packages not images. Given that history there will always be "exceptions" when building in obs. I considered running udev on the worker which would have prevented that patch. But doing so requires a package that starts the service. So all you do in a worker is aligned to a package. This is the fundamental concept of obs and will not change.

I would go with your suggestion to use a variable as it seems upstream supports this. But the problem is really that all the distros adds their special sauce and this can't be trusted as general solution. From that background I rather maintain a little special sauce in our own scope which we also cover by tests.

Does this makes sense ? I think we can maintain it but I don't want to add something you are not convinced with.

Thanks

Reply:
> Does this makes sense ? I think we can maintain it but I don't want to add something you are not convinced with.

Yes it does. We can certainly maintain it. Any other alternative that I can think to avoid those corner case considerations relay on quite uncommon use of grub2 tools.

Reply:
ok, thanks. I'll give it a try right now and if it proves to fix the issue I'll merge this and also submit to Factory. Once done I'll update the sle12-sp5 branch with the patch.